api/client/samples/analogous_years/lib/get_transform_data.py

```diff
@@ -28,6 +28,8 @@ def get_data(client, metric_id, item_id, region_id, source_id, frequency_id, sta
                    'start_date': start_date}
     client.add_single_data_series(data_series)
     data = client.get_df()
+    # TODO: select data from dataframe such that the date start_date restrictions are maintained
+    #  and the client does not need to be re-initiated in a single jupyter notebook
```

Comment:
1. The answer to the first question - The reason I did not take this route initially is - I thought I am not changing the data type of the variable `start_date` to datetime until later, which is not the case. So, should I revert back all the changes and just incorporate this?

2. I think this is separate from your first question. If I understand you correctly your concern is about the following lines 
 ```
 if data['end_date'].iloc[0] > start_date:
        new_value = data['value'].iloc[0]
        new_row = pd.DataFrame({'end_date': [start_date], 'value': [new_value]})
        data = pd.concat([new_row, data[:]]).reset_index(drop=True)
```
I had to put it there so that there is a common start date for all the series. Otherwise, the function `loop_start_date` was not producing desired result. Example: If I add rainfall along with ENSO, I was not getting the correct `loop_start_date`.

Reply:
Discussed with Yiqing. Keeping the previously done changes; but also incorporating the code Nemo suggested. The idea is passing `access_token` for the client instead of `client` may be a better idea.

Reply:
re: 1, thx looks good.

re: 2... yes that was separate point. I understand you are padding it to make it line up with start date. My concern is 
a)  if the first data point p is such that  p.start_date <= start_date  < p.end_date, thenthe partially covered p.value is overweighted instead of under.  For example if it's a weekly series and the data points start,end and value are like:
1/11, 1/17, v1
1/18,1/24,v2
and now you get a start_date of 1/20. This would return 
1/20,v2
1/24,v2
This is imprecise as, the true value for the period ending 1/20 is closer to v1 than v2.

b) More importantly if the data starts later, i.e.  start_date < p.start_date <= p.end_date, say start_date is 1/5 in the above example, then you are returning
1/5,v1
1/11,v1
but it would be more accurate to retrun
1/5,null
1/11,v1

To fix both you could change the downstream logic to deal with data points as representing a value over an interval of time ((start_date, end_date), value) and deal with boundaries as needed. But that would eb a bigger change than this PR.  

For now, i think you shuold at least fix scenario b).  e.g. also keep start_date column and 
```
if data['end_date'].iloc[0] > start_date:
    if data['start_date'].iloc[0] <= start_date:
          new_value = null
   else:
          ....   
```

Reply:
Your comments are spot on in terms of initial point being unfairly weighed. 

There is at least one other situation where the first point is weighed more. The case when the data series is not sorted by date. John assured me that does not happen with Gro data series. A situation like that would arise if I stuck with this code and fixed the failing unit_test. We did consider this and arrived at the decision that 1 data point weighed against 365 data point would not be enough to change the rank monumentally. Hence, I kept this code. 

Obviously, there are cases where I would be completely wrong - For example the worst scenario in a data series with half yearly frequency. or even monthly frequency. That is why there is a disclaimer that this works best for daily frequency data series.

Regarding your points about p.start_date < p.end_date. I investigated three examples -
`series_enso = {'metric_id': 15851977,
                        'item_id': 13495,
                        'region_id': 0,
                        'source_id': 124,
                        'frequency_id': 6}`
`series_ym = {
	'metric_id': 170037, 
	'item_id': 274, 
	'region_id': 1215, 
	'source_id': 32, 
	'frequency_id': 9, 
	'unit_id': 284,
  'show_revisions': True
}`

`series_1 ={
	'metric_id': 15820065, 
	'item_id': 270, 
	'region_id': 1215, 
	'partner_region_id': 0, 
	'source_id': 81, 
	'frequency_id': 15, 
	'show_revisions': True
}`

and none of them had a different p.start_date and p.end_date. I was under the impression that would be the case in general. Is that not true?

Reply:
> but it would be more accurate to retrun
> 1/5,null
> 1/11,v1

The way it is set up, analogous_years need to have all the data series at the initial and final date with non-null value (computing euclidean distance with null value would not be possible). So, I do not think null would work.

Reply:
Discussed offline: there needs to be a separate PR to make sure that I do not drop the series start_date values. As they are valuable information in case of monthly, semi-annual frequency.

Reply:
I stuck with the previously discussed done in this PR, with a TODO for the next PR. I look towards almighty reviewers for approval!!